sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,570 @@
+    available_read_locations: Vec<String>,
+    available_write_endpoints_by_location: HashMap<String, String>,
+    available_read_endpoints_by_location: HashMap<String, String>,
+    write_endpoints: Vec<String>,
+    read_endpoints: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct LocationUnavailabilityInfo {
+    pub last_check_time: SystemTime,
+    pub unavailable_operation: RequestOperation,
+}
+
+pub struct LocationCache {
+    pub default_endpoint: String,
+    pub locations_info: DatabaseAccountLocationsInfo,
+    pub location_unavailability_info_map: Mutex<HashMap<String, LocationUnavailabilityInfo>>,
+}
+
+impl LocationCache {
+    pub fn new(default_endpoint: String, preferred_locations: Vec<String>) -> Self {
+        Self {
+            default_endpoint,
+            locations_info: DatabaseAccountLocationsInfo {
+                preferred_locations: preferred_locations,
+                ..Default::default()
+            },
+            location_unavailability_info_map: Mutex::new(HashMap::new()),
+        }
+    }
+
+    pub fn update(
+        &mut self,
+        write_locations: HashMap<String, String>,
+        read_locations: HashMap<String, String>,
+    ) -> Result<(), &'static str> {
+        // Separate write locations into appropriate hashmap and list
+        if !write_locations.is_empty() {
+            let (available_write_endpoints_by_location, available_write_locations) =
+                self.get_endpoints_by_location(write_locations);
+            self.locations_info.available_write_endpoints_by_location =
+                available_write_endpoints_by_location;
+            self.locations_info.available_write_locations = available_write_locations;
+        }
+
+        // Separate read locations into appropriate hashmap and list
+        if !read_locations.is_empty() {
+            let (available_read_endpoints_by_location, available_read_locations) =
+                self.get_endpoints_by_location(read_locations);
+            self.locations_info.available_read_endpoints_by_location =
+                available_read_endpoints_by_location;
+            self.locations_info.available_read_locations = available_read_locations;
+        }
+
+        self.refresh_endpoints();
+
+        Ok(())
+    }
+
+    pub fn mark_endpoint_unavailable(&mut self, endpoint: &str, operation: RequestOperation) {
+        let now = std::time::SystemTime::now();
+
+        {
+            let mut location_unavailability_info_map =
+                self.location_unavailability_info_map.lock().unwrap();
+
+            if let Some(info) = location_unavailability_info_map.get_mut(endpoint) {
+                if info.unavailable_operation == operation
+                    || info.unavailable_operation == RequestOperation::All
+                {
+                    // If the endpoint is already marked as unavailable, update last_check_time
+                    info.last_check_time = now;
```

Comment:
Since you do this in both branches, you could just move it to before the if on line 99